fix: Only include project LICENSE and NOTICE in Spark Client Jar

RussellSpitzer opened the pull request:
Previously the build code would add LICENSE and NOTICE files from other shaded dependencies, some of which could potentially shade the project's LICENSE or NOTICE or cause confusion as to which file was correct.

To fix this a post-processing task is added which removes all potentially confusing files and manually adds the project LICENSE and NOTICE files back to the jar.

---

I tried fixing this via exclusions and include rules but for whatever reason the shadowJar plugin didn't seem to behave
well. The Duplication strategy apparently can only apply to the final jar so unless I was able to individually exclude the
licenses from all dependencies I was a bit stuck. To avoid the issue entirely, I just post process the jar. This adds about 10 
seconds to the task on my machine but I figure that is not a huge cost for something that won't be run that often.

RussellSpitzer commented:
Output from task -
```
➜  polaris git:(CleanupSparkPluginLicense) ✗ ./gradlew :polaris-spark-3.5_2.12:clean :polaris-spark-3.5_2.12:createPolarisSparkJar
Configuration on demand is an incubating feature.

> Task :polaris-spark-3.5_2.12:addLicenseFilesToJar
Processing jar: /Users/rspitzer/repos/polaris/plugins/spark/v3.5/spark/build/2.12/libs/polaris-spark-3.5_2.12-1.1.0-incubating-SNAPSHOT-bundle.jar
Using temp directory: /Users/rspitzer/repos/polaris/plugins/spark/v3.5/spark/build/2.12/tmp/jar-cleanup-polaris-spark-3.5_2.12-bundle
Removing license file: LICENSE-EDL-1.0.txt
Removing license file: LICENSE-EPL-1.0.txt
Removing license file: LICENSE
Removing license file: META-INF/LICENSE
Removing license file: META-INF/NOTICE
Removing license file: META-INF/LICENSE.txt
Removing license file: NOTICE
Added project LICENSE file
Added project NOTICE file
Recreated jar with only project LICENSE and NOTICE files

[Incubating] Problems report is available at: file:///Users/rspitzer/repos/polaris/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/8.14.2/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 14s
23 actionable tasks: 2 executed, 3 from cache, 18 up-to-date
➜  polaris git:(CleanupSparkPluginLicense) ✗ jar tf ./plugins/spark/v3.5/spark/build/2.12/libs/polaris-spark-3.5_2.12-1.1.0-incubating-SNAPSHOT-bundle.jar | grep -iE "LICENSE|NOTICE"
META-INF/licenses/
META-INF/licenses/org.projectnessie.nessie/
META-INF/licenses/org.projectnessie.nessie/nessie-client-0.103.3/
META-INF/licenses/org.projectnessie.nessie/nessie-model-0.103.3/
LICENSE
META-INF/FastDoubleParser-LICENSE
META-INF/FastDoubleParser-NOTICE
META-INF/bigint-LICENSE
META-INF/licenses/org.projectnessie.nessie/nessie-client-0.103.3/LICENSE
META-INF/licenses/org.projectnessie.nessie/nessie-client-0.103.3/NOTICE
META-INF/licenses/org.projectnessie.nessie/nessie-model-0.103.3/LICENSE
META-INF/licenses/org.projectnessie.nessie/nessie-model-0.103.3/NOTICE
META-INF/thirdparty-LICENSE
NOTICE
codegen/includes/license.ftl
org/eclipse/microprofile/openapi/annotations/info/License.class
org/eclipse/microprofile/openapi/models/info/License.class
```

gh-yzou commented on plugins/spark/v3.5/spark/build.gradle.kts:
```diff
@@ -99,9 +93,91 @@ tasks.register<ShadowJar>("createPolarisSparkJar") {
+        include("META-INF/NOTICE*")
+      }
+      .forEach { file ->
+        println("Removing license file: ${file.relativeTo(tempDir)}")
```
shall we remove the print statement in the build?

RussellSpitzer replied:
On second thought, Gradle DEBUG is very noisy, i'll put it at INFO

gh-yzou reviewed (approved):
@RussellSpitzer Thanks for helping on the license for spark client!

flyrain commented on plugins/spark/v3.5/spark/build.gradle.kts:
```diff
@@ -99,9 +93,91 @@ tasks.register<ShadowJar>("createPolarisSparkJar") {
+      .matching {
+        include("LICENSE*")
+        include("NOTICE*")
+        include("META-INF/LICENSE*")
+        include("META-INF/NOTICE*")
```
I guess we could do this 
```
   include("META-INF/*LICENSE")
        include("META-INF/*NOTICE")
```

dimas-b commented on plugins/spark/v3.5/spark/build.gradle.kts:
```diff
@@ -99,9 +93,91 @@ tasks.register<ShadowJar>("createPolarisSparkJar") {
+
+// Post-processing task to add our project's LICENSE and NOTICE files to the jar and remove any
+// other LICENSE or NOTICE files that were shaded in.
+tasks.register("addLicenseFilesToJar") {
```
suggestion: Would this work?
1. Run `shadowJar` with `exclude("LICENSE")`
2. Run a simple jar task depending on `shadowJar` and add our own `LICENSE`
3. Use output of step 3 as the published artifact.

It doubles the fat jar, but hopefully it's not too much overhead.

RussellSpitzer replied:
Nope :( I tried that first

RussellSpitzer replied:
Quick demonstration -

```
diff --git a/plugins/spark/v3.5/spark/build.gradle.kts b/plugins/spark/v3.5/spark/build.gradle.kts
index 58ac8c98..6a9defb3 100644
--- a/plugins/spark/v3.5/spark/build.gradle.kts
+++ b/plugins/spark/v3.5/spark/build.gradle.kts
@@ -89,6 +89,9 @@ tasks.register<ShadowJar>("createPolarisSparkJar") {
   from(sourceSets.main.get().output)
   configurations = listOf(project.configurations.runtimeClasspath.get())

+  exclude("LICENSE")
+  exclude("NOTICE")
+
   // Optimization: Minimize the JAR (remove unused classes from dependencies)
   // The iceberg-spark-runtime plugin is always packaged along with our polaris-spark plugin,
   // therefore excluded from the optimization.
   ```
   
   ```
   > Task :polaris-spark-3.5_2.12:addLicenseFilesToJar
Custom actions are attached to task ':polaris-spark-3.5_2.12:addLicenseFilesToJar'.
Caching disabled for task ':polaris-spark-3.5_2.12:addLicenseFilesToJar' because:
  Gradle would require more information to cache this task
Task ':polaris-spark-3.5_2.12:addLicenseFilesToJar' is not up-to-date because:
  Task has not declared any outputs despite executing actions.
Processing jar: /Users/rspitzer/repos/polaris/plugins/spark/v3.5/spark/build/2.12/libs/polaris-spark-3.5_2.12-1.1.0-incubating-SNAPSHOT-bundle.jar
Using temp directory: /Users/rspitzer/repos/polaris/plugins/spark/v3.5/spark/build/2.12/tmp/jar-cleanup-polaris-spark-3.5_2.12-bundle
Removing license file: LICENSE-EDL-1.0.txt
Removing license file: LICENSE-EPL-1.0.txt
Removing license file: LICENSE. <------------------Not Excluded :_
Removing license file: META-INF/LICENSE
Removing license file: META-INF/NOTICE
Removing license file: META-INF/LICENSE.txt
Removing license file: NOTICE . <------------------Not Excluded :_
Added project LICENSE file
Added project NOTICE file
[ant:jar] Building jar: /Users/rspitzer/repos/polaris/plugins/spark/v3.5/spark/build/2.12/libs/polaris-spark-3.5_2.12-1.1.0-incubating-SNAPSHOT-bundle.jar
Recreated jar with only project LICENSE and NOTICE files
```

dimas-b commented:
Ah, I guess what I mentioned in my previous comment has already been addressed :facepalm:

RussellSpitzer commented:
I got worried about the codegen license template, so I decided to not touch lower-case license files.

```
Removing license file: LICENSE-EDL-1.0.txt
Removing license file: LICENSE-EPL-1.0.txt
Removing license file: LICENSE
Removing license file: META-INF/FastDoubleParser-NOTICE
Removing license file: META-INF/LICENSE
Removing license file: META-INF/licenses/org.projectnessie.nessie/nessie-client-0.103.3/LICENSE
Removing license file: META-INF/licenses/org.projectnessie.nessie/nessie-client-0.103.3/NOTICE
Removing license file: META-INF/licenses/org.projectnessie.nessie/nessie-model-0.103.3/LICENSE
Removing license file: META-INF/licenses/org.projectnessie.nessie/nessie-model-0.103.3/NOTICE
Removing license file: META-INF/FastDoubleParser-LICENSE
Removing license file: META-INF/thirdparty-LICENSE
Removing license file: META-INF/bigint-LICENSE
Removing license file: META-INF/NOTICE
Removing license file: META-INF/LICENSE.txt
Removing license file: NOTICE
Removed META-INF/licenses directory
Added project LICENSE file
Added project NOTICE file
```

jbonofre reviewed (approved):
The exclude is good (it's OK from this standpoint).

However, the `LICENSE` has to be updated to document Google Flatbuffers.
